src/library/dao/trackdao.cpp

```diff
@@ -1545,143 +1548,162 @@ bool TrackDAO::detectMovedTracks(QSet<TrackId>* pTracksMovedSetOld,
+                    << oldTrackLocation;
+            continue;
         }
 
-        //If we found a moved track...
-        if (newTrackLocationId.isValid()) {
-            qDebug() << "Found moved track!" << filename;
-
-            // Remove old row from track_locations table
-            query.prepare("DELETE FROM track_locations WHERE id=:id");
-            query.bindValue(":id", oldTrackLocationId.toVariant());
-            if (!query.exec()) {
-                // Should not happen!
-                LOG_FAILED_QUERY(query);
-            }
+        TrackId oldTrackId(oldTrackQuery.value(oldTrackIdColumn));
+        DEBUG_ASSERT(oldTrackId.isValid());
+        DbId oldTrackLocationId(oldTrackQuery.value(oldLocationIdColumn));
+        DEBUG_ASSERT(oldTrackLocationId.isValid());
 
-            // The library scanner will have added a new row to the Library
-            // table which corresponds to the track in the new location. We need
-            // to remove that so we don't end up with two rows in the library
-            // table for the same track.
-            query.prepare("SELECT id FROM library WHERE location=:location");
-            query.bindValue(":location", newTrackLocationId.toVariant());
-            if (!query.exec()) {
-                // Should not happen!
-                LOG_FAILED_QUERY(query);
-            }
+        // The queries ensure that this assertion is always true (fs_deleted=0/1)!
+        DEBUG_ASSERT(oldTrackId != newTrackId);
 
-            const int query3idColumn = query.record().indexOf("id");
-            if (query.next()) {
-                TrackId newTrackId(query.value(query3idColumn));
-                query.prepare("DELETE FROM library WHERE id=:newid");
-                query.bindValue(":newid", newTrackLocationId.toVariant());
-                if (!query.exec()) {
-                    // Should not happen!
-                    LOG_FAILED_QUERY(query);
-                }
-                // We collect all the new tracks the where added to BaseTrackCache as well
-                pTracksMovedSetNew->insert(newTrackId);
-            }
-            // Delete the track
+        // The library scanner will have added a new row to the Library
+        // table which corresponds to the track in the new location. We need
+        // to remove that so we don't end up with two rows in the library
+        // table for the same track.
+        {
+            QSqlQuery query(m_database);
             query.prepare("DELETE FROM library WHERE id=:newid");
-            query.bindValue(":newid", newTrackLocationId.toVariant());
-            if (!query.exec()) {
-                // Should not happen!
+            query.bindValue(":newid", newTrackId.toVariant());
+            VERIFY_OR_DEBUG_ASSERT(query.exec()) {
                 LOG_FAILED_QUERY(query);
             }
-
+        }
+        if (oldTrackLocationId != newTrackLocationId) {
```

Comment:
What is the condition that this is false? 
Please add a code comment.

Reply:
I will replace it with a debug assertion